### PR TITLE
Add stream support for response body - piped-io-stream impl

### PR DESCRIPTION
**Pull request by pete23**
Following patch implements streaming support as per discussion in https://github.com/ring-clojure/ring-json/issues/58 - apologies for the delay.

Note that streaming is configured for the whole handler - it's either all or nothing. Testing for a lazy sequence to enable the stream doesn't make sense in retrospect - I might have a large on-heap fully realised structure I can't create an on heap string for.

The default is no streaming; the wrapper needs to specify { :stream true } as an option to enable this functionality so I think your request that this be unobtrusive and not change any defaults has been honoured.

Added a couple of functional tests - let me know if you would like more coverage.

In terms of non-functional tests,

```clojure
(defn core-handler [req] {:status 200 :headers {"Content-Type" "text/plain"} :body (repeat 100000000 {:a "hello" :b "world"})})
(def handler (wrap-json-response core-handler {:stream true :pretty true}))
```

This test crashes after just over a minute with java.lang.OutOfMemoryError: Java heap space with stream = false and default Java 12 JVM settings. 

With stream = true it completes generation of 3.5Gb of data (at a rate of around 21.7Mb/s) with heap size fluctuating under 100Mb as garbage is generated and collected.

### CODE REVIEW

File: src/ring/middleware/json.clj [hunk 1]
```diff
@@ -100,12 +102,19 @@
        (handler request respond raise)
        (respond malformed-response)))))
 
+(defn- json-stream
+  "Adapts the output stream from Cheshire to the input stream required by Ring."
```
Comment (weavejester) [inline src/ring/middleware/json.clj]: Can you remove this docstring? It's not a public API so it doesn't need one.

File: src/ring/middleware/json.clj [hunk 2]
```diff
@@ -118,7 +127,8 @@
   Accepts the following options:
 
   :pretty            - true if the JSON should be pretty-printed
-  :escape-non-ascii  - true if non-ASCII characters should be escaped with \\u"
+  :escape-non-ascii  - true if non-ASCII characters should be escaped with \\u
+  :stream            - true if JSON body should be streamed rather than a string"
```
Comment (weavejester) [inline src/ring/middleware/json.clj]: Change this key to: `:stream?`.

File: src/ring/middleware/json.clj [hunk 3]
```diff
@@ -100,12 +102,19 @@
...
+   (fn [out] (json/generate-stream body (writer out) options))))
+
 (defn json-response
   "Converts responses with a map or a vector for a body into a JSON response.
   See: wrap-json-response."
   [response options]
   (if (coll? (:body response))
-    (let [json-resp (update-in response [:body] json/generate-string options)]
+    (let [generator (if (:stream options) json-stream json/generate-string)
+          json-resp (update-in response [:body] generator options)]
```
Comment (weavejester) [inline src/ring/middleware/json.clj]: `:stream` keyword shouldn't be passed to generator.

File: test/ring/middleware/test/json.clj [hunk 4]
```diff
@@ -337,6 +337,20 @@
       (is (= (get-in response [:headers "Content-Type"]) "application/json; some-param=some-value"))
       (is (= (:body response) "{\"foo\":\"bar\"}")))))
 
+(deftest test-json-response-streaming
+  (testing "streaming vector body"
+    (let [handler  (constantly {:status 200 :headers {} :body [:foo :bar]})
+          response ((wrap-json-response handler {:stream true}) {})]
+      (is (= (get-in response [:headers "Content-Type"]) "application/json; charset=utf-8"))
+      (is (= (slurp (:body response)) "[\"foo\",\"bar\"]")))) ;; using slurp checks that we have an inputstream
```
Comment (weavejester) [inline test/ring/middleware/test/json.clj]: No need for the comment here, as I think it's self-evident what `slurp` is being used for.

File: src/ring/middleware/json.clj [hunk 5]
```diff
@@ -2,6 +2,8 @@
   "Ring middleware for parsing JSON requests and generating JSON responses."
   (:require [cheshire.core :as json]
             [cheshire.parse :as parse]
+            [clojure.java.io :refer [writer]]
```
Comment (weavejester) [inline src/ring/middleware/json.clj]: Prefer `:as io` here.

File: src/ring/middleware/json.clj [hunk 6]
```diff
@@ -2,6 +2,8 @@
...
+            [ring.util.io :refer [piped-input-stream]]
```
Comment (weavejester) [inline src/ring/middleware/json.clj]: Prefer `:as ring-io`.

File: src/ring/middleware/json.clj (hunk 2 above)
Comment (pete23) [reply src/ring/middleware/json.clj]: No problem, was just staying consistent with other opts.

File: src/ring/middleware/json.clj (hunk 3 above)
Comment (pete23) [reply src/ring/middleware/json.clj]: Laziness. Apologies.

File: src/ring/middleware/json.clj (hunk 1 above)
Comment (pete23) [reply src/ring/middleware/json.clj]: i’ll change the method name then to better express intent.

**Comment by pete23** [comment]
OK think have resolved everything. Also updated the versions tested so I can confirm that the version I use in anger is ok.

Apologies if there any stupid PR newbie issues, this is my first time doing this on GH.

File: src/ring/middleware/json.clj (hunk 2 above)
Comment (weavejester) [reply src/ring/middleware/json.clj]: Yep, that's understandable. Those options in particular are passed through to Cheshire, which is why they're named that way, rather than being consistent with the rest of the code.

File: src/ring/middleware/json.clj [hunk 7]
```diff
@@ -100,12 +102,18 @@
        (handler request respond raise)
        (respond malformed-response)))))
 
+(defn- generate-input-stream
+  [body options]
```
Comment (weavejester) [inline src/ring/middleware/json.clj]: The arg list can be put on one the same like as the `defn-`

File: project.clj [hunk 8]
```diff
@@ -7,8 +7,10 @@
                  [cheshire "5.7.1"]
                  [ring/ring-core "1.6.2"]]
   :plugins [[codox "0.8.13"]]
-  :aliases {"test-all" ["with-profile" "default:+1.6:+1.7:+1.8" "test"]}
+  :aliases {"test-all" ["with-profile" "default:+1.6:+1.7:+1.8:+1.9:+1.10" "test"]}
```
Comment (weavejester) [inline project.clj]: These changes need to be separated out into their own PR, as they're not directly related to adding stream support.

File: src/ring/middleware/json.clj [hunk 9]
```diff
@@ -100,12 +102,18 @@
...
+  (ring-io/piped-input-stream
+   (fn [out] (json/generate-stream body (io/writer out) options))))
+
 (defn json-response
   "Converts responses with a map or a vector for a body into a JSON response.
   See: wrap-json-response."
-  [response options]
+  [response {:keys [stream?] :as options}]
```
Comment (weavejester) [inline src/ring/middleware/json.clj]: Destructuing in the arg list affects the documentation generator, so it either needs to be overridden with metadata, or just use `(:stream? options)` instead.

File: src/ring/middleware/json.clj [hunk 10]
```diff
@@ -100,12 +102,18 @@
...
   (if (coll? (:body response))
-    (let [json-resp (update-in response [:body] json/generate-string options)]
+    (let [generator (if stream? generate-input-stream json/generate-string)
+          json-resp (update-in response [:body] generator (dissoc options :stream?))]
```
Comment (weavejester) [inline src/ring/middleware/json.clj]: Please ensure lines are within 80 characters. I have some [contributing guidelines](https://github.com/weavejester/contributing/blob/master/CONTRIBUTING.md) that I've yet to add to all my projects, especially the older ones like Ring-JSON.

File: project.clj (hunk 8 above)
Comment (pete23) [reply project.clj]: removed and will resubmit separately

**Comment by weavejester** [comment]
Sorry for taking so long to respond to this. I've been working my way through a sizeable backlog of PRs.

This PR uses `piped-input-stream`, but past Ring 1.6 there's a better way to achieve the same result. I'm debating whether to use `piped-input-stream` for compatibility, or to use `StreamableResponseBody`, or to check for the presence of the protocol and use one method or the other. I'm leaning toward the latter, but I may merge the PR and make the changes in a later commit.

**Comment by pete23** [comment]
ring-json seems to have a dep on ring 1.6.2 anyway. Do you want me to recode this patch for the StreamableResponseBody protocol instead?

Runtime checking and providing both implementations seems like gilding the lily to me but obv your call.

**Comment by weavejester** [comment]
Let's use the `StreamableResponseBody` protocol, and if anyone complains we can add extra implementations :)

**Comment by pete23** [comment]
Actually having read https://www.booleanknot.com/blog/2016/07/15/asynchronous-ring.html three times now I'm still not sure what the StreamableResponseBody version would look like. Either:

1) Extend the protocol for the types processed by wrap-json-response to stream out JSON, and hence obviate the whole middleware? ring-json currently picks up anything that gives true to coll so that would be pretty fundamental. Maybe not.

2) Have our own helper class that extends the protocol, wraps and then JSON serialises the body content.

So 2 then. Update the request to have a JSONResponseBody that implements StreamableResponseBody protocol, holds the original body, and does the Right Thing when called wrt streaming, and that this behaviour should be behind an optional flag in the same way as the current patch.

I'm out of steam for this evening and might not get a chance to look at this for a couple of days. I'll start from scratch with a separate PR when I do - I'd like to leave this here for now as it applies cleanly and does the job, in case anyone needs the job doing before we get it right:-)

**Comment by weavejester** [comment]
Right, I was thinking 2:

```clojure
(defrecord JsonStreamingResponseBody [body options]
  ring.core.protocols/StreamableResponseBody
  (write-body-to-stream [_ _ output-stream]
    (json/generate-stream body (io/writer output-stream) options)))
```

And then replace `generate-input-stream` with `->JsonStreamingResponseBody`.

This avoids having to pipe through an intermediate stream - we can just write directly to the output stream that's exposed by the web server.

**Comment by pete23** [comment]
OK, this is where my head was at by the end of my previous post too. I’ll find time tomorrow to apply, test, and ship another PR.

**Comment by pete23** [comment]
Ok the better version is in #60 now, I'll close this off.

**Comment by weavejester** [comment]
In general I prefer that changes are made to the existing PR rather than a new one created. There's no advantage to creating a new PR; it just makes it harder to read the thread of comments and code reviews.

**Comment by pete23** [comment]
I know, but I’d managed to get my branch in an awful state.

**Comment by pete23** [comment]
As the other version was a different implementation entirely I thought it warranted a separate PR.
